feat: Swift export is in Alpha (#5560)
JetBrains/kotlin-web-site: docs/topics/native/native-swift-export.md

Reviewer: That's fixed.

  generated Swift code.
* **Module name customization**. You can customize the resulting Swift module names in the Gradle configuration of your
  Kotlin project.

> * **Concurrency support**. You can seamlessly call suspending Kotlin code from Swift and export `kotlinx.coroutines`'

Reviewer 2: ```suggestion
* **Concurrency support**. You can seamlessly call suspending Kotlin code from Swift and export `kotlinx.coroutines`
```

Reviewer 2: The possessive while imho correct looks really strange here, kotlinx.coroutines flows works I think

>   flows out of the box into Swift's `AsyncSequence`.

Reviewer 2: ```suggestion
  flows as Swift's `AsyncSequence` out of the box.
```

Reviewer 2: (suggestion) not sure if it works, but it reads clearer to me

## Enable Swift export

> Swift export is currently in [Alpha](components-stability.md#stability-levels-explained) and is still incomplete, so breaking changes are expected.

Reviewer 2: ```suggestion
Swift export is currently in [Alpha](components-stability.md#stability-levels-explained) and still incomplete, so breaking changes are expected.
```

Reviewer 2: (suggestion) reads a bit more natural without it, (is currently , is still)

The table below shows how Kotlin concepts are mapped to Swift.

| Kotlin                                     | Swift                          |
|--------------------------------------------|--------------------------------|
| [`class`](#classes)                        | `class`                        |
| [`object`](#objects)                       | `class` with `shared` property |
| [`enum class`](#enums)                     | `enum`                         |
| [`typealias`](#type-aliases)               | `typealias`                    |
| [Function](#functions)                     | Function                       |
| [`suspend fun`](#suspending-functions)     | `async`                        |

> | [`kotlinx.coroutines`' flows](#flow-types) | `AsyncSequence`                |

Reviewer 2: ```suggestion
| [`kotlinx.coroutines` flows](#flows) | `AsyncSequence`                |
```

public func log(messages: Swift.String...)
```

> > Support for functions with `inline` and `operator` keywords is currently limited.

Reviewer: > `inline` 

We have added the support for those also :)

}
```

### Concurrency

#### Suspending functions

> You can call suspending Kotlin code from Swift. Kotlin [`suspend` functions](composing-suspending-functions.md) and

Reviewer 2: ```suggestion
You can call suspending Kotlin code from Swift. Kotlin [suspending functions](coroutines-basics.md#suspending-functions) and
```

Reviewer 2: it's a more up to date description so I'd suggest linking to this one

> suspend functional types are exported as Swift's `async` counterparts:

Reviewer 2: ```suggestion
suspending functional types are exported as Swift's `async` counterparts:
```

```kotlin
// Kotlin
suspend fun hello(): String {
    delay(1000)
    return "Hello Swift! This is Kotlin."
}
```

```swift
// Swift
let msg = try await hello()
```

> #### Flow types

Reviewer 2: (suggestions) While not wrong, I'd suggest Flows in the title to better match the previous one — Suspending functions

> You can also export `kotlinx.coroutines`' flows as Swift's [`AsyncSequence`](https://developer.apple.com/documentation/Swift/AsyncSequence):

Reviewer 2: ```suggestion
You can also export `kotlinx.coroutines` flows as Swift's [`AsyncSequence`](https://developer.apple.com/documentation/Swift/AsyncSequence):
```

Reviewer 2: maybe we could link to flows as well - but it can wait until we publish the changes there 🤔

```kotlin
// Kotlin

> // Type String is preserved when exporting Flow

Reviewer 2: ```suggestion
// Preserves the String type when exporting Flow
```

fun flowOfStrings(): Flow<String> = flowOf("hello", "any", "world")
```

```swift
// Swift
var actual: [String] = []

> // Type String is correctly inferred from Kotlin

Reviewer 2: I think the word "correctly" is a bit redundant so I'd suggest dropping it.

> // Type String is correctly inferred from Kotlin

Reviewer 2: ```suggestion
// Infers the String type from Kotlin
```

for try await element in flowOfStrings().asAsyncSequence() {
    actual.append(element)
}
```

> #### Dispatcher management

Reviewer 2: Dispatcher management reads a bit strange compared to the other titles, I think we could keep this similar to the rest: 
Coroutine dispatchers

By default, when calling a Kotlin `suspend` function from Swift, as well as when using `asAsyncSequence`, a coroutine

> context utilizing the `Default` dispatcher is created on the Kotlin side, and execution is delegated to it.

Reviewer 2: I'd suggest something like this for these two lines: 

By default, when you call a Kotlin suspending function from Swift or use the `.asAsyncSequence()` function, Kotlin runs the exported code in a coroutine context that uses the [`Dispatchers.Default`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html) dispatcher.

> If you want to execute Kotlin code on a different dispatcher, you switch the context manually in Kotlin:

Reviewer 2: ```suggestion
To run the exported code on a different dispatcher, use `withContext()` to switch the coroutine context in Kotlin:
```

Reviewer 2: feels a bit more specific I think we should definitely mention `withContext()` in the text as well

Reviewer 2: we could even link to https://kotlinlang.org/docs/coroutines-basics.html#coroutine-dispatchers for more information

Reviewer 2: or perhaps: 

To run the exported code in a coroutine context that uses a different dispatcher, use the `withContext()` function:

#### Suspending functions

You can call suspending Kotlin code from Swift. Kotlin [suspending functions](coroutines-basics.md#suspending-functions) and
suspending functional types are exported as Swift's `async` counterparts:

```kotlin
// Kotlin
suspend fun hello(): String {
    delay(1000)
    return "Hello Swift! This is Kotlin."
}
```

```swift
// Swift
let msg = try await hello()
```

#### Flows

You can also export `kotlinx.coroutines` flows as Swift's [`AsyncSequence`](https://developer.apple.com/documentation/Swift/AsyncSequence):

```kotlin
// Kotlin
// Preserves the String type when exporting Flow
fun flowOfStrings(): Flow<String> = flowOf("hello", "any", "world")
```

```swift
// Swift
var actual: [String] = []

// Infers the String type from Kotlin
for try await element in flowOfStrings().asAsyncSequence() {
    actual.append(element)
}
```

#### Coroutine dispatchers

By default, when you call a Kotlin suspending function from Swift or use the `asAsyncSequence` function,
Kotlin creates a coroutine context that uses the [`Dispatchers.Default`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html)

> dispatcher and executes the exported code there.

Reviewer 2: nit:
My only worry with this phrasing is that it might be ambiguous what we mean by "there" here? perhaps:  and executes the exported code in that coroutine context. — but that might make it too long.. 🤔
